pkg/features/kube_features.go

```diff
@@ -699,6 +699,13 @@ const (
 	// certificate as expiration approaches.
 	RotateKubeletServerCertificate featuregate.Feature = "RotateKubeletServerCertificate"
 
+	// owner: @kiashok
+	// kep: https://kep.k8s.io/4216
+	// alpha: v1.29
+	//
+	// Adds support to pull images based on the runtime class specified
+	RuntimeClassInImageCriApi featuregate.Feature = "RuntimeClassInImageCriApi"
```

Comment:
Should this be added in isolation without the code where it is being used? We don't want a situation where the feature gate merges but rest of the code misses the release.

Reply:
Yea I would add this with the code you want to be guarded by the feature.

Reply:
yeah suggest draft the pr till get the code and any nec. test improvements..

Reply:
I've updated the PR with the necessary changes under the feature gate. I am not sure adding tests would make sense yet as no changes have been implemented on the container runtime side (example in containerd) and we cannot validate through functional or e2e tests in k8s. I have validated it against a private draft version of the containerd changes on AKS cluster.

Reply:
see https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/images/image_manager_test.go#L338

Reply:
also in the e2e node checks it would be ok to write the test bucket and add "skips" for the container runtime releases that are not ready yet.. can do manual testing if not available in any.. and the skips get removed when support avail and the gate is on

Reply:
Hmm I am not sure we would be able to test anything at all meaningfully with the fakeruntimes. I feel it would make more sense to add these tests once some functionality is in container runtimes like containerd atleast. As of now all this is doing is to read runtimehandler field that is already in podspec and setting ImageSpec.Runtimehandler.

Any thoughts on this @aravindhp @jsturtevant @marosset @haircommander ?

Reply:
I think adding unit tests to cover the feature toggling logic would be very important.  You won't really cover those in e2e tests.  

I think you can add tests pretty easily to all these points.  You don't need to have a e2e test but it at least covers the case of the feature toggle off/on and logic makes sense.

Reply:
@kannon92 https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4216-image-pull-per-runtime-class#can-the-feature-be-disabled-once-it-has-been-enabled-ie-can-we-roll-back-the-enablement turning feature off and testing needs manual steps as mentioned in the KEP. I am not sure that can be tested? We could test to ensure ListImages returns with the runtimeHandler of the image set etc

Reply:
Yea for E2E I understand that.  I still think you could test `ListImages` with feature toggle.  Same for things like convert.go, image_manager, etc.  You actually should try to test those components with unit as it would be difficult to cover those areas with e2e tests.  E2E tests can be used to verify the full implementation including container runtime but there is no reason why that should stop you from writing unit tests with the fake CRI.

Reply:
@mikebrow updated PR to include test. Please let me know if this looks good. Thanks.

Reply:
Updated PR with all unit tests for functionality added under the new feature gate